pkg/server/sandbox_run.go

```diff
@@ -77,55 +73,34 @@ func (c *criContainerdService) RunPodSandbox(ctx context.Context, r *runtime.Run
 	// Ensure sandbox container image snapshot.
 	image, err := c.ensureImageExists(ctx, c.sandboxImage)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get sandbox image %q: %v", defaultSandboxImage, err)
-	}
-	rootfsMounts, err := c.snapshotService.View(ctx, id, image.ChainID)
-	if err != nil {
-		return nil, fmt.Errorf("failed to prepare sandbox rootfs %q: %v", image.ChainID, err)
-	}
-	defer func() {
-		if retErr != nil {
-			if err := c.snapshotService.Remove(ctx, id); err != nil {
-				glog.Errorf("Failed to remove sandbox container snapshot %q: %v", id, err)
-			}
-		}
-	}()
-	var rootfs []*types.Mount
-	for _, m := range rootfsMounts {
-		rootfs = append(rootfs, &types.Mount{
-			Type:    m.Type,
-			Source:  m.Source,
-			Options: m.Options,
-		})
+		return nil, fmt.Errorf("failed to get sandbox image %q: %v", c.sandboxImage, err)
 	}
 
 	// Create sandbox container.
 	spec, err := c.generateSandboxContainerSpec(id, config, image.Config)
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate sandbox container spec: %v", err)
 	}
-	rawSpec, err := json.Marshal(spec)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal oci spec %+v: %v", spec, err)
-	}
+
 	glog.V(4).Infof("Sandbox container spec: %+v", spec)
-	if _, err = c.containerService.Create(ctx, containers.Container{
-		ID: id,
-		// TODO(random-liu): Checkpoint metadata into container labels.
-		Image:   image.ID,
-		Runtime: containers.RuntimeInfo{Name: defaultRuntime},
-		Spec: &prototypes.Any{
-			TypeUrl: runtimespec.Version,
-			Value:   rawSpec,
-		},
-		RootFS: id,
-	}); err != nil {
+
+	// TODO(random-liu): Checkpoint metadata into container labels.
+	var opts []containerd.NewContainerOpts
+	opts = append(opts, []containerd.NewContainerOpts{
+		containerd.WithSpec(spec),
+		containerd.WithImage(image.Image),
+		containerd.WithRuntime(defaultRuntime),
+		containerd.WithNewSnapshotView(id, image.Image)}...)
```

Comment:
There is a problem with the containerd client. If `NewContainer` fails, no one will cleanup the snapshot.

In our original implementation, we'll remove the snapshot. But in their client, it's not removed.

This is not your problem, will file an issue to them. But could you add a TODO here to track that?